File: Java/Plasma/PubUtilLib/plGImage/plBitmap.java
package Plasma.PubUtilLib.plGImage;

import Plasma.NucleusLib.pnKeyedObject.*;
import Plasma.CoreLib.*;
import java.io.IOException;

public class plBitmap extends hsKeyedObject {
    public static final byte IMG_ARGB = 0;
    public static final byte IMG_DXT  = 1;
    public static final byte IMG_JPEG = 2;

    protected byte version;
    protected byte BPP, CPB;
    protected short flags;
    protected byte imgType;
    protected byte texelSize, texelType;
    protected short Unk1, Unk2;

    public plBitmap() { }

    public void read(hsStream S) throws IOException {
        super.read(S);

        version = S.readByte();
        BPP = S.readByte();
        CPB = S.readByte();
        flags = S.readShort();
        imgType = S.readByte();
        if (imgType != 0 && imgType != IMG_JPEG)
            texelSize = S.readByte();
        else texelSize = 0;
        texelType = S.readByte();
        Unk1 = S.readShort();
        Unk2 = S.readShort();
    }

    public void write(hsStream S) throws IOException {
        super.write(S);

        S.writeByte(version);
        S.writeByte(BPP);
        S.writeByte(CPB);
        S.writeShort(flags);
        S.writeByte(imgType);
        if (imgType != 0 && imgType != IMG_JPEG)
            S.writeByte(texelSize);
        S.writeByte(texelType);
        S.writeShort(Unk1);
        S.writeShort(Unk2);
    }

    public int getTypeNum() { return 3; }
    public String getTypeName() { return "plBitmap"; }
}

File: C++/PubUtilLib/plGImage/plBitmap.cpp
#include "plBitmap.h"

plBitmap::plBitmap(PlasmaVer pv) : BPP(0), CPB(0), flags(0), imgType(0),
                       texelSize(0), texelType(0), Unk1(0), Unk2(0) { }

plBitmap::~plBitmap() { }

void plBitmap::read(hsStream * S) {
    hsKeyedObject::read(S);
    readData(S);
}

void plBitmap::readData(hsStream * S) {
    S->readByte();  // Version == 2
    BPP = S->readByte();
    CPB = S->readByte();
    flags = S->readShort();
    imgType = S->readByte();
    if ((imgType != IMG_ARGB) && (imgType != IMG_JPEG))
        texelSize = S->readByte();
    else texelSize = 0;
    texelType = S->readByte();
    Unk1 = S->readInt();
    Unk2 = S->readInt();
}

void plBitmap::write(hsStream * S) {
    hsKeyedObject::write(S);
    writeData(S);
}

void plBitmap::writeData(hsStream * S) {
    S->writeByte(BITMAPVER);
    S->writeByte(BPP);
    S->writeByte(CPB);
    S->writeShort(flags);
    S->writeByte(imgType);
    if ((imgType != IMG_ARGB) && (imgType != IMG_JPEG))
        S->writeByte(texelSize);
    S->writeByte(texelType);
    S->writeInt(Unk1);
    S->writeInt(Unk2);
}
